rr_evgp/src/binary_bayes_filter_obstacle_layer.cpp

```diff
@@ -0,0 +1,250 @@
+
+        for (int mx = min_cell_x; mx < max_cell_x; mx++) {
+            for (int my = min_cell_y; my < max_cell_y; my++) {
+                updates_[master_grid.getIndex(mx, my)] = -1;
+            }
+        }
+
+        double angle = scan_->angle_min;
+        for (double raw_range : scan_->ranges) {
+            bool is_hit = (!std::isinf(raw_range) && raw_range <= scan_range_);
+            double range = is_hit ? raw_range : scan_range_;
+            double trace_range = range + (resolution * 0.7);
+            double wx = lidar_x_ + trace_range * std::cos(lidar_yaw_ + angle);
+            double wy = lidar_y_ + trace_range * std::sin(lidar_yaw_ + angle);
+
+            if (wx < world_min_x || wx >= world_max_x || wy < world_min_y || wy >= world_max_y) {
+                ROS_WARN_THROTTLE(1.0, "trying to scan outside the map");
+                continue;
+            }
+
+            const double step_size = resolution * 0.25;
+            const double step_x = step_size * (wx - lidar_x_) / range;
+            const double step_y = step_size * (wy - lidar_y_) / range;
+            int mx, my;
+            double d = 0;
+            double march_x = lidar_x_;
+            double march_y = lidar_y_;
+            const double last_clear = range - resolution;
+            while (d < last_clear) {
+                master_grid.worldToMapNoBounds(march_x, march_y, mx, my);
+                updates_[master_grid.getIndex(mx, my)] = 0;
+                d += step_size;
+                march_x += step_x;
+                march_y += step_y;
+            }
+            if (is_hit) {
+                while (d <= trace_range) {
+                    master_grid.worldToMapNoBounds(march_x, march_y, mx, my);
+                    updates_[master_grid.getIndex(mx, my)] = 1;
+                    d += step_size;
+                    march_x += step_x;
+                    march_y += step_y;
+                }
+            }
+
+            angle += scan_->angle_increment;
+        }
+
+        uint8_t* costmap_data = master_grid.getCharMap();
+        for (int mx = min_cell_x; mx < max_cell_x; mx++) {
+            for (int my = min_cell_y; my < max_cell_y; my++) {
+                const auto i = master_grid.getIndex(mx, my);
+                const char update = updates_[i];
+
+                if (update >= 0) {
+                    double p_prior = probs_[i];
+                    if (update > 0) {  // hit
+                        double p_occupied = (1.0 - prob_false_neg_) * p_prior;
+                        double p_empty = prob_false_pos_ * (1.0 - p_prior);
+                        probs_[i] = std::min(p_occupied / (p_occupied + p_empty), max_prob_);
+                    } else {  // cleared
+                        double p_occupied = prob_false_neg_ * p_prior;
+                        double p_empty = (1.0 - prob_false_pos_) * (1.0 - p_prior);
+                        probs_[i] = std::max(p_occupied / (p_occupied + p_empty), 1.0 - max_prob_);
```

Comment:
no log odds notation? I think it'll also be more numerically stable

Reply:
Log odds would be a good future improvement